content-src/lib/selectLayoutRender.js

```diff
@@ -94,4 +106,28 @@ export const selectLayoutRender = (state, prefs, rickRollCache) => {
       return {...component, data};
     }),
   })).filter(row => row.components.length);
+
+  // Generate the payload for the SPOCS Fill ping. Note that a SPOC could be rejected
+  // by the `probability_selection` first, then gets chosen for the next position. For
+  // all other SPOCS that never went through the probabilistic selection, its reason will
+  // be "out_of_position".
+  let spocsFill = [];
+  if (spocs.data.spocs) {
```

Comment:
@ncloudioj  spocs.data.spocs will throw error if spocs.data is undefined, can we add a check for that edge case.  thanks!

Reply:
Fixed, thank! Did you see this error happen? The init state of `spocs.data` was set as `{}` in the reducer, I wonder how/when would it be `undefined`?

Reply:
yes, came across this error if a faulty layout_endpoint is used in discoverystream.config, which fails to fetch endpoint on the first load

Reply:
Ah, I see. We might want to change that so the fetch failures won't affect the state structure. Otherwise, we'd have to check this all over the place.

Reply:
That will be the correct fix, it's [this call ](https://github.com/mozilla/activity-stream/blob/master/lib/DiscoveryStreamFeed.jsm#L370) that sets spocs to undefined

Reply:
Filed bug https://bugzilla.mozilla.org/show_bug.cgi?id=1548929